saliency/xrai.py

```diff
@@ -0,0 +1,467 @@
+        del remaining_masks[key]
+      if len(remaining_masks) == 0:
+        break
+      added_mask = remaining_masks[best_key]
+      mask_diff = _get_diff_mask(added_mask, current_mask)
+      masks_trace.append((mask_diff, best_gain))
+
+      current_mask = np.logical_or(current_mask, added_mask)
+      current_attr_sum = np.sum(attr[current_mask])
+      current_area_perc = np.mean(current_mask)
+      output_attr[mask_diff] = best_gain
+      del remaining_masks[best_key]  # delete used key
+      if verbose:
+        logging.info(
+            "{} of {} masks added,"
+            "attr_sum: {}, area: {:.3g}/{:.3g}, {} remaining masks".format(
+                added_masks_cnt, n_masks, current_attr_sum, current_area_perc,
+                area_perc_th, len(remaining_masks)))
+      added_masks_cnt += 1
+
+    uncomputed_mask = output_attr == -np.inf
+    # Assign the uncomputed areas a value such that sum is same as ig
+    output_attr[uncomputed_mask] = gain_fun(uncomputed_mask, attr)
+    # Set uncomputed region's rank to max rank + 1
+    masks_trace = zip(*sorted(masks_trace, key=lambda x: -x[1]))[0]
+    if integer_segments:
+      attr_ranks = np.zeros(shape=attr.shape, dtype=np.int)
+      for i, mask in enumerate(masks_trace):
+        attr_ranks[mask] = i + 1
+      assert np.all(uncomputed_mask == (attr_ranks == 0))
+      attr_ranks[uncomputed_mask] = np.max(attr_ranks) + 1
+      return output_attr, attr_ranks
+    else:
+      return output_attr, masks_trace
+
+  @staticmethod
+  def _xrai_fast(attr,
+                 segs,
+                 gain_fun=_gain_density,
+                 area_perc_th=1.0,
+                 verbose=0,
+                 integer_segments=True):
+    """We expect attr to be 2D, XRAI shape is equal to attr shape
+      Segs are list of binary masks, one per segment
+    """
+    output_attr = -np.inf * np.ones(shape=attr.shape, dtype=np.float)
+
+    n_masks = len(segs)
+    current_mask = np.zeros(attr.shape, dtype=bool)
+
+    masks_trace = []
+
+    # Sort all masks based on gain, ignore overlaps
+    seg_attrs = [gain_fun(seg_mask, attr) for seg_mask in segs]
+    segs, seg_attrs = zip(
+        *sorted(zip(segs, seg_attrs), key=lambda x: -x[1]))
+
+    for i, added_mask in enumerate(segs):
+      mask_diff = _get_diff_mask(added_mask, current_mask)
+      mask_gain = gain_fun(mask_diff, attr)
+      masks_trace.append((mask_diff, mask_gain))
+      output_attr[mask_diff] = mask_gain
+      current_mask = np.logical_or(current_mask, added_mask)
+      if verbose:
+        current_attr_sum = np.sum(attr[current_mask])
+        current_area_perc = np.mean(current_mask)
+        logging.info("{} of {} masks added,"
+                     "attr_sum: {}, area: {:.3g}/{:.3g}".format(
+                         i + 1, n_masks, current_attr_sum, current_area_perc,
+                         area_perc_th))
+    masks_trace = zip(*sorted(masks_trace, key=lambda x: -x[1]))[0]
```

Comment:
'zip' in Python 3 returns an iterable, so it cannot be indexed directly. Use list(zip(...))[0] instead.

Reply:
done.